File: hcx-pipeline-jobs/preauth-job/src/main/scala/org/swasth/dp/preauth/task/PreauthStreamTask.scala
package org.swasth.dp.preauth.task

import java.io.File
import java.util
import com.typesafe.config.ConfigFactory
import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.api.java.typeutils.TypeExtractor
import org.apache.flink.api.java.utils.ParameterTool
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment
import org.swasth.dp.core.function.ContextEnrichmentFunction
import org.swasth.dp.core.job.FlinkKafkaConnector
import org.swasth.dp.core.util.FlinkUtil
import org.swasth.dp.preauth.functions.PreauthProcessFunction

class PreauthStreamTask(config: PreauthConfig, kafkaConnector: FlinkKafkaConnector) {

  private val serialVersionUID = 146697324640926024L

  def process(): Unit = {
    implicit val env: StreamExecutionEnvironment = FlinkUtil.getExecutionContext(config)
    implicit val mapTypeInfo: TypeInformation[util.Map[String, AnyRef]] = TypeExtractor.getForClass(classOf[util.Map[String, AnyRef]])
    implicit val stringTypeInfo: TypeInformation[String] = TypeExtractor.getForClass(classOf[String])
    val kafkaConsumer = kafkaConnector.kafkaMapSource(config.kafkaInputTopic)
    val enrichedStream =
      env.addSource(kafkaConsumer, config.eligibilityCheckConsumer)
        .uid(config.eligibilityCheckConsumer).setParallelism(config.kafkaConsumerParallelism)
        .rebalance()
        .process(new ContextEnrichmentFunction(config)).setParallelism(config.downstreamOperatorsParallelism)

    val eventStream = enrichedStream.getSideOutput(config.enrichedOutputTag)
        .process(new PreauthProcessFunction(config)).setParallelism(config.downstreamOperatorsParallelism)

    /** Sink for retry events */
    eventStream.getSideOutput(config.retryOutputTag).addSink(kafkaConnector.kafkaMapSink(config.retryTopic)).name(config.retryProducer).uid(config.retryProducer).setParallelism(config.downstreamOperatorsParallelism)

    /** Sink for audit events */
    eventStream.getSideOutput(config.auditOutputTag).addSink(kafkaConnector.kafkaStringSink(config.auditTopic)).name(config.auditProducer).uid(config.auditProducer).setParallelism(config.downstreamOperatorsParallelism)

    env.execute(config.jobName)
  }
}

// $COVERAGE-OFF$ Disabling scoverage as the below code can only be invoked within flink cluster
object PreauthStreamTask {
  def main(args: Array[String]): Unit = {
    val configFilePath = Option(ParameterTool.fromArgs(args).get("config.file.path"))
    val config = configFilePath.map {
      path => ConfigFactory.parseFile(new File(path)).resolve()
    }.getOrElse(ConfigFactory.load("preauth.conf").withFallback(ConfigFactory.systemEnvironment()))
    val pipelinePreprocessorConfig = new PreauthConfig(config)
    val kafkaUtil = new FlinkKafkaConnector(pipelinePreprocessorConfig)
    val task = new PreauthStreamTask(pipelinePreprocessorConfig, kafkaUtil)
    task.process()
  }
}

// $COVERAGE-ON$

File: hcx-pipeline-jobs/preauth-job/src/main/java/org/swasth/dp/preauth/task/PreauthStreamTask.java
package org.swasth.dp.preauth.task;

import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.java.typeutils.TypeExtractor;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.connector.kafka.source.KafkaSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.swasth.dp.core.function.ContextEnrichmentFunction;
import org.swasth.dp.core.job.BaseJobConfig;
import org.swasth.dp.core.job.FlinkKafkaConnector;
import org.swasth.dp.core.util.FlinkUtil;
import org.swasth.dp.preauth.functions.PreauthProcessFunction;
import scala.Option;
import scala.Some;

import java.io.File;
import java.util.Map;

public class PreauthStreamTask {

  private PreauthConfig config;
  private FlinkKafkaConnector kafkaConnector;
  public PreauthStreamTask(PreauthConfig config, FlinkKafkaConnector kafkaConnector){
    this.config = config;
    this.kafkaConnector = kafkaConnector;
  }

  public static void main(String[] args) {
    Option<String> configFilePath = new Some<String>(ParameterTool.fromArgs(args).get("config.file.path"));
    Config conf = configFilePath.map(path -> ConfigFactory.parseFile(new File(path)).resolve())
            .getOrElse(() -> ConfigFactory.load("resources/preauth.conf").withFallback(ConfigFactory.systemEnvironment()));
    PreauthConfig config = new PreauthConfig(conf,"Preauth-Job");
    FlinkKafkaConnector kafkaConnector = new FlinkKafkaConnector(config);
    PreauthStreamTask task = new PreauthStreamTask(config, kafkaConnector);
    try {
      task.process(config);
    } catch (Exception e) {
      //TODO Add loggers
      e.printStackTrace();
    }
  }

  void process(BaseJobConfig baseJobConfig) throws Exception {
    StreamExecutionEnvironment env = FlinkUtil.getExecutionContext(baseJobConfig);
    KafkaSource<Map<String, Object>> kafkaConsumer = kafkaConnector.kafkaMapSource(config.kafkaInputTopic);

    SingleOutputStreamOperator<Map<String,Object>> enrichedStream = env.fromSource(kafkaConsumer, WatermarkStrategy.noWatermarks(), config.preauthConsumer)
            .uid(config.preauthConsumer).setParallelism(config.consumerParallelism)
            .rebalance()
            .process(new ContextEnrichmentFunction(config, TypeExtractor.getForClass(String.class))).setParallelism(config.downstreamOperatorsParallelism);

    SingleOutputStreamOperator<Map<String,Object>> eventStream = enrichedStream.getSideOutput(config.enrichedOutputTag())
            .process(new PreauthProcessFunction(config)).setParallelism(config.downstreamOperatorsParallelism);

    /** Sink for audit events */
    eventStream.getSideOutput(config.auditOutputTag()).sinkTo(kafkaConnector.kafkaStringSink(config.auditTopic()))
            .name(config.auditProducer()).uid(config.auditProducer()).setParallelism(config.downstreamOperatorsParallelism);

    System.out.println(config.jobName() + " is processing");
    env.execute(config.jobName());
  }

}
